src/amp-story-embed.js

```diff
@@ -123,34 +194,145 @@ export class AmpStoryEmbed {
   }
 
   /**
-   * @return {!Promise}
    * @public
    */
   layoutCallback() {
-    // TODO: Layout all child iframes.
-    return this.layoutIframe_(0);
+    if (this.isLaidOut_) {
+      return;
+    }
+
+    const iframes = toArray(this.rootEl_.querySelectorAll('iframe'));
+    // TODO(Enriqe): Layout all child iframes.
+    this.layoutIframe_(this.stories_[0], iframes[0]);
+
+    this.isLaidOut_ = true;
   }
 
   /**
-   * @param {number} index
-   * @return {!Promise}
+   * @param {!Element} story
+   * @param {!Element} iframe
    * @private
    */
-  layoutIframe_(index) {
-    const story = this.stories_[index];
-    this.iframeEl_.setAttribute('src', story.href);
-    return Promise.resolve();
+  layoutIframe_(story, iframe) {
+    const {href} = this.getEncodedUrl_(story.href);
+
+    iframe.setAttribute('src', href);
+  }
+
+  /**
+   * Gets encoded url for viewer usage.
+   * @param {string} href
+   * @return {!Location}
+   * @private
+   */
+  getEncodedUrl_(href) {
+    const {location} = this.win_;
+    const url = parseUrlWithA(this.cachedA_, location.href);
+
+    const params = dict({
+      'amp_js_v': '0.1',
+      'visibilityState': 'inactive',
+      'origin': url.origin,
+    });
+
+    let inputUrl = addParamsToUrl(href, params);
+    inputUrl = inputUrl.replace('amp_js_v=0.1', 'amp_js_v=0.1#');
```

Comment:
Isn't this going to leave an extra `&` after the `#`?

Also, this won't work as `addParamsToUrl` seems to be adding the query parameter before the `hash`. You'd break the canonical URL by having the original hash at the very end of the URL.

Reply:
Good catch! I am now extracting the original fragment param first before prepending it again after adding the new parameters.